graphql/authentication.ts

```diff
@@ -246,4 +272,52 @@ export class AuthenticationResolver {
 
         return true;
     }
+
+    @Authorized(Role.UNAUTHENTICATED)
+    @Mutation((returns) => SSOAuthStatus)
+    async loginWithSSO(@Ctx() context: GraphQLContext, @Arg('code') code: string, @Arg('referrer') referrer: string) {
+        const { email, firstname, lastname, clientId, sub } = await authenticateWithIDP({ code, referrer });
+        if (!email || !firstname) {
+            throw new Error('Invalid token payload: Missing required fields (email/name)');
+        }
+
+        const setSSORegisteringUser = async () => {
+            const newRoles = new Set(context.user.roles.concat(Role.SSO_REGISTERING_USER));
+            await userSessions.set(context.sessionToken, {
+                ...context.user,
+                roles: Array.from(newRoles),
+                email,
+                firstname,
+                lastname,
+                idpClientId: clientId,
+                idpSub: sub,
+            });
+        };
+
+        let userId: string;
+        try {
+            // Get our userId using IDP sub/clientId
+            userId = await getUserIdFromIDPLogin(sub, clientId);
+        } catch (error) {
+            // If there is no credentials for that, check if the email is already taken.
+            if (!(await isEmailAvailable(email))) {
+                // If so, we'll try to link that email with the given sub/clientId
+                await setSSORegisteringUser();
+                return SSOAuthStatus.link;
+            }
+        }
+
+        let user: User;
+        try {
+            user = await getUser(userId, true);
+        } catch (error) {
```

Comment:
I don't follow, how can `getUser(...)` ever fail? Shouldn't this be the `else` branch of isEmailAvailable?